### DISCUSSION TITLE
'Type level' inference of the natural type of a lambda

### DISCUSSION THREAD

**Discussion by MCGPPeters** [dotnet/csharplang#6809] [Language Ideas] [answered]
With the lamba improvements in c# 10 (https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/lambda-improvements)  it is now possible to infer the type of a lambda:

`var foo = (string s) => (string t) => $"You passed the string {s + t}";` // Func<string, Func<string, string>>

This makes it very natural to write curried function definitions. Would it be possible to make this available on the type level as well? Now I would have to write:

`public Func<string, Func<string, string>> foo = (string s) => (string t) => $"You passed the string {s + t}";`

Which is very clunky... Maybe something like:

`public function Foo = (string s) => (string t) => $"You passed the string {s + t}";` ? 

Taking this train of thought a little further:

`static abstract function Foo = (string s) => (string t) => string` ?

**Comment by CyrusNajmabadi** [comment, marked as answer]
It's definitely a reasonable ask, but one we've explicitly rejected in the past (in the same vein as allowing 'var' at the type-level) because it would make determining the type of something dependent on binding the expression within.  But binding the expression within needs that type-level-information.  This creates a circular dependency in type-information that we feel is too detrimental to our language and compiler design.

**Reply by MCGPPeters** [reply]
Ok.... Too bad. Thanks for the fast response on this. Clear.
